File: paddler_client_javascript/src/schemas/AgentIssue.ts
import { z } from "zod";

import { AgentIssueModelPathSchema } from "./AgentIssueModelPath";
import { HuggingFaceDownloadLockSchema } from "./HuggingFaceDownloadLock";

export const AgentIssueSchema = z.union([
  z.object({
    CacheCannotAcquireLock: AgentIssueModelPathSchema,
  }),
  z.object({
    CacheDirectoryIsNotWritable: AgentIssueModelPathSchema,
  }),
  z.object({
    CacheStorageIsFull: AgentIssueModelPathSchema,
  }),
  z.object({
    ChatTemplateDoesNotCompile: z.object({
      error: z.string(),
      model_path: AgentIssueModelPathSchema,
      template_content: z.string(),
    }),
  }),
  z.object({
    DownloadServerDeniedAccess: AgentIssueModelPathSchema,
  }),
  z.object({
    DownloadServerErrored: AgentIssueModelPathSchema,
  }),
  z.object({
    DownloadServerIsUnreachable: AgentIssueModelPathSchema,
  }),
  z.object({
    DownloadServerRejectedRequest: AgentIssueModelPathSchema,
  }),
  z.object({
    DownloadInterrupted: AgentIssueModelPathSchema,
  }),
  z.object({
    DownloadUrlIsMalformed: AgentIssueModelPathSchema,
  }),
  z.object({
    HuggingFaceCannotAcquireLock: HuggingFaceDownloadLockSchema,
  }),
  z.object({
    HuggingFaceModelDoesNotExist: AgentIssueModelPathSchema,
  }),
  z.object({
    HuggingFacePermissions: AgentIssueModelPathSchema,
  }),
  z.object({
    ModelCacheIsCorrupted: AgentIssueModelPathSchema,
  }),
  z.object({
    ModelCannotBeLoaded: AgentIssueModelPathSchema,
  }),
  z.object({
    ModelDoesNotExistAtUrl: AgentIssueModelPathSchema,
  }),
  z.object({
    ModelFileDoesNotExist: AgentIssueModelPathSchema,
  }),
  z.object({
    MultimodalProjectionCannotBeLoaded: AgentIssueModelPathSchema,
  }),
  z.object({
    SlotCannotStart: z.object({
      error: z.string(),
      slot_index: z.number(),
    }),
  }),
  z.object({
    UnableToFindChatTemplate: AgentIssueModelPathSchema,
  }),
]);

export type AgentIssue = z.infer<typeof AgentIssueSchema>;

File: paddler_messaging/src/agent_issue.rs
use serde::Deserialize;
use serde::Serialize;

use crate::agent_issue_params::chat_template_does_not_compile_params::ChatTemplateDoesNotCompileParams;
use crate::agent_issue_params::hugging_face_download_lock::HuggingFaceDownloadLock;
use crate::agent_issue_params::model_path::ModelPath;
use crate::agent_issue_params::slot_cannot_start_params::SlotCannotStartParams;

#[derive(Clone, Debug, Deserialize, Eq, Hash, Ord, PartialEq, PartialOrd, Serialize)]
#[serde(deny_unknown_fields)]
pub enum AgentIssue {
    CacheCannotAcquireLock(ModelPath),
    CacheDirectoryIsNotWritable(ModelPath),
    CacheStorageIsFull(ModelPath),
    ChatTemplateDoesNotCompile(ChatTemplateDoesNotCompileParams),
    DownloadInterrupted(ModelPath),
    DownloadServerDeniedAccess(ModelPath),
    DownloadServerErrored(ModelPath),
    DownloadServerIsUnreachable(ModelPath),
    DownloadServerRejectedRequest(ModelPath),
    DownloadUrlIsMalformed(ModelPath),
    HuggingFaceCannotAcquireLock(HuggingFaceDownloadLock),
    HuggingFaceModelDoesNotExist(ModelPath),
    HuggingFacePermissions(ModelPath),
    ModelCacheIsCorrupted(ModelPath),
    ModelCannotBeLoaded(ModelPath),
    ModelDoesNotExistAtUrl(ModelPath),
    ModelFileDoesNotExist(ModelPath),
    MultimodalProjectionCannotBeLoaded(ModelPath),
    SlotCannotStart(SlotCannotStartParams),
    UnableToFindChatTemplate(ModelPath),
}
